ably/realtime_channel_spec_test.go

```diff
@@ -155,7 +155,804 @@ func attachAndWait(t *testing.T, channel *ably.RealtimeChannel) {
+		defer safeclose(t, close)
+
+		close = c.To(
+			connecting,
+			connected,
+		)
+
+		defer safeclose(t, close)
+
+		channelTransitioner := c.Channel("test")
+		channel := channelTransitioner.Channel
+
+		channelStateChanges := make(ably.ChannelStateChanges, 10)
+		channel.OnAll(channelStateChanges.Receive)
+
+		channelTransitioner.ToSpecifiedState(chAttaching)
+
+		var channelStatechange ably.ChannelStateChange
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		var rg ablytest.ResultGroup
+		defer rg.Wait()
+		rg.GoAdd(func(ctx context.Context) error {
+			err := channel.Detach(ctx)
+			if err != nil {
+				t.Fatal(err)
+			}
+			return err
+		})
+
+		ablytest.Instantly.NoRecv(t, nil, channelStateChanges, t.Fatalf) // Shouldn't send detach waiting to get attached
+
+		channelTransitioner.ToSpecifiedState(chAttached)
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateAttached, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Instantly.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateDetaching, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+
+		ablytest.Soon.Recv(t, &channelStatechange, channelStateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateDetached, channelStatechange.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, channelStatechange)
+		}
+	})
+
+	// Suspend not implemented, so directly setting the state
+	t.Run("RTL5j: if channel state is SUSPENDED, immediately transition to DETACHED state", func(t *testing.T) {
+		t.Parallel()
+
+		_, _, _, channel, stateChanges, _ := setup(t)
+		ctx, cancel := context.WithCancel(context.Background())
+		cancel()
+		channel.OnAll(stateChanges.Receive)
+
+		channel.SetState(ably.ChannelStateSuspended, nil)
+		ablytest.Instantly.Recv(t, nil, stateChanges, t.Fatalf) // State will be changed to suspended
+
+		channel.Detach(ctx)
+
+		var change ably.ChannelStateChange
+		ablytest.Instantly.Recv(t, &change, stateChanges, t.Fatalf)
+		if expected, got := ably.ChannelStateDetached, change.Current; expected != got {
+			t.Fatalf("expected %v; got %v (event: %+v)", expected, got, change)
+		}
+		ablytest.Instantly.NoRecv(t, &change, stateChanges, t.Fatalf)
+	})
+
+	// Hard to reproduce, because channel ideally should be in attaching when attach is received, and channel detach waits for attaching state
+	t.Run("RTL5k: When receive ATTACH in detaching state, send new DETACH message", func(t *testing.T) {
```

Comment:
@tcard This one as per the spec. I wasn't able to find any reference implementation from the Dotnet client. Same as suspend, this one can't be produced via transitioner or custom dials.

Reply:
Why not? The test controls the incoming messages with the `in` channel. Why not just move to DETACHING normally, and at that point send an ATTACHED, see that a new DETACH is sent on `out`?

Reply:
Although we can reproduce it locally, it seems highly unlikely in real case. For this to happen, ably needs to send attached message, it can only send it when local state is in attaching state and attach message is sent ... even if times out and state transitions to suspended, for detaching to be set by detach .. it has to be in attached state, for suspended it will just return detached directly. I cant test it, but his is a very hypothetical case I feel ... there are good chances of this to happen when ably servers start tripping and send random attach messages,..

Reply:
It is a corner case, but a well-defined (and well-behaved) one that needs to be supported. The server may send an ATTACHED at any point ([RTL12](https://docs.ably.io/client-lib-development-guide/features/#RTL12), no need to be trippin' :) ), and it just may happen that it crosses paths with a DETACH from the client.

In any case, it's in the spec, so it needs to be supported.

Reply:
Yeah, seems I missed it out. So, yes, this is a valid use case.

Reply:
Fixed !